src/semipoly.jl

```diff
@@ -3,225 +3,304 @@ using DataStructures
+# its degree is less than or equal to `degree_bound`.
+# If `m` is a constant about `vars`, return true if `consts = true` and return false if
+# `consts = false`.
+function isboundedmonomial(m, vars, degree_bound::Real; consts = true)::Bool
+    if !(m isa SemiMonomial)
+        return false
+    end
+    degree_dict = pdegrees(m.p)
+    if isempty(degree_dict)
+        return consts && !has_vars(m.coeff, vars)
+    end
+    degrees = values(degree_dict)
+    for degree in degrees
+        if !isinteger(degree) || degree < 0
+            return false
+        end
+    end
+    if sum(degrees) > degree_bound
+        return false
+    end
+    !has_vars(m.coeff, vars)
+end
+
+# Return true if the degrees of `m` are all 0s and its coefficient is a `Real`.
+Base.:isreal(m::SemiMonomial) = m.p isa Number && isone(m.p) && unwrap(m.coeff) isa Real
+Base.:isreal(::Symbolic) = false
 
-_degree(x::BoundedDegreeMonomial) = x.overdegree ? Inf : pdegree(x.p)
+# Transform `m` to a `Real`.
+# Assume `isreal(m) == true`, otherwise calling this function does not make sense.
+function Base.:real(m::SemiMonomial)::Real
+    if isinteger(m.coeff)
+        return Int(m.coeff)
+    end
+    return m.coeff
+end
 
-_degree(x) = isop(x, *) ? sum(_degree, unsorted_arguments(x)) : 0
+symtype(m::SemiMonomial) = symtype(m.p)
 
+TermInterface.issym(::SemiMonomial) = true
 
-## Semi-polynomial form
+Base.:nameof(m::SemiMonomial) = Symbol(:SemiMonomial, m.p, m.coeff)
 
-isboundedmonom(x) = x isa BoundedDegreeMonomial && !x.overdegree
-bareterm(x, f, args;kw...) = Term{symtype(x)}(f, args)
+isop(x, op) = istree(x) && operation(x) === op
+isop(op) = Base.Fix2(isop, op)
 
-function mark_and_exponentiate(expr, vars, deg, consts)
+bareterm(x, f, args; kw...) = Term{symtype(x)}(f, args)
 
+function mark_and_exponentiate(expr, vars)
     # Step 1
     # Mark all the interesting variables -- substitute without recursing into nl forms
-
-    expr′ = mark_vars(expr, vars, consts)
+    expr′ = mark_vars(expr, vars)
 
     # Step 2
-    # Construct and propagate BoundedDegreeMonomial for ^ and *
-
-
-    rules = [@rule (~a::isboundedmonom) ^ (~b::(x-> x isa Integer && x > 0)) =>
-             BoundedDegreeMonomial(((~a).p)^(~b), (~a).coeff ^ (~b), !_isone((~a).p) && ~b > deg)
-             @rule (~a::isop(+)) ^ (~b::(x -> x isa Integer)) => pow_of_add(~a, ~b, deg, vars, consts)
-
-             @rule *(~~x) => mul_bounded(~~x, deg)]
-
-    expr′ = Postwalk(RestartedChain(rules), similarterm=bareterm)(expr′)
+    # Construct and propagate BoundedDegreeMonomial for ^ and * and /
+
+    # does not do fraction simplification
+    rules = [@rule (~a::issemimonomial)^(~b::isreal) => (~a)^real(~b)
+             @rule (~a::isop(+))^(~b::isreal) => expand(Pow((~a), real(~b)))
+             @rule *(~~xs::(xs -> all(issemimonomial, xs))) => *(~~xs...)
+             @rule *(~~xs::(xs -> any(isop(+), xs))) => expand(Term(*, ~~xs))
```

Comment:
I avoided expand here so that something like `(1+x)^1000` upto degree 2 would not take up a lot of time and memory... Can you benchmark this vs your branch vs and the previous one? It maybe nice to bring it back.

Reply:
Yes. We should bring it back. But the current partial multinominal expansion doesn't work in some edge cases. 

And it's weird that the CI with Julia v1.7.3 takes 13 minutes, but the CI with Julia v1.6.7 takes more than 1 hour.

Reply:
Are you going to try to do this in this PR or should I merge it and wait for another PR?

Reply:
Please merge. I'll open another PR later this week. Thank you.